README.md
# STAT 545B Assignment B1 and B2

Project B1 and B2 for STAT 545B

1. What is this repository?

    This repository contains my Assignments B1 and B2 for the course `STAT545B`. These assignments cover making a function in R, documenting it, and testing it, along with evaluating my ability to wrap the function created in B1 (or something else!) into an R package.


   
2. What files are in this GitHub repository?
   
   *There is currently* ***1*** *file, along with* ***2***  *folders in my GitHub repository*
   
   ```
    - README.md
    - Assignment B1
    - vancomycin.calculator
   ```
   
3. How do you use this repository?

    Assuming you are taking STAT545B, you can either create a new project using this repository as a directory or you can download the files at your discretion. If you opted to create a new project, setup your local environment and create a pull request to obtain all the files. Alternatively, just download the files of interest to you. You can also look through each individual folder to explore the contents. Each folder has a dedicated README, so refer to that specific README when trying to understand what was done. Ideally, you should begin with the Assignment B1 folder. Once you have finished with the Assignment B1 folder, you can install the function in the form of a package from the vancomycin.calculator folder.

vancomycin.calculator/NEWS.md
# vancomycin.calculator 0.1.0

* Package can be used to calculate the individualized vancomycin dosage for a patient
* Update to the output of the function and unit correction

# vancomycin.calculator 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


vancomycin.calculator/tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(vancomycin.calculator)

test_check("vancomycin.calculator")


vancomycin.calculator/tests/testthat/test-Vanco_SS_dose_calculator.R
Vanco_SS_1 <- digest::digest(Vanco_SS_dose_calculator(11.0, 22.1, 1000, 12, 1.5, 1.33333, 55/60, 35, 17.5)) # digesting the Vanco_SS_dose_calculator input
Vanco_SS_2 <- digest::digest(Vanco_SS_dose_calculator(25.0, 33.5, 1500, 8, 1.5, 115/60, 92/60, 33.5, 17.5)) # digesting the Vanco_SS_dose_calculator input
Vanco_SS_3 <- digest::digest(Vanco_SS_dose_calculator(c(25,25), c(33.5, 43.0), c(1500,1000), c(8,8), c(1.5,1), c(115/60,1.5), c(92/60, 1), c(33.5,35), c(17.5:17.5))) # digesting the Vanco_SS_dose_calculator input

test_that("Testing Vanco_SS_dose_calculator function design",{
  expect_equal(digest::digest(Vanco_SS_dose_calculator(11.0, 22.1, 1000, 12, 1.5, 1.33333, 55/60, 35, 17.5)), Vanco_SS_1)
}) # expect_equal test comparing the input of Vanco_SS_dose_calculator(11.0, 22.1, 1000, 12, 1.5, 1.33333, 55/60, 35, 17.5) to the digested input

test_that("Testing Vanco_SS_dose_calculator function design",{
  expect_equal(digest::digest(Vanco_SS_dose_calculator(25.0, 33.5, 1500, 8, 1.5, 115/60, 92/60, 33.5, 17.5)), Vanco_SS_2)
}) # expect_equal test comparing the input of Vanco_SS_dose_calculator(25.0, 33.5, 1500, 8, 1.5, 115/60, 92/60, 33.5, 17.5) to the digested input

test_that("Testing Vanco_SS_dose_calculator function design",{
  expect_equal(digest::digest(Vanco_SS_dose_calculator(c(25,25), c(33.5, 43.0), c(1500,1000), c(8,8), c(1.5,1), c(115/60,1.5), c(92/60, 1), c(33.5,35), c(17.5:17.5))), Vanco_SS_3)
}) # expect_equal test comparing the input of Vanco_SS_dose_calculator(c(25,25), c(33.5, 43.0), c(1500,1000), c(8,8), c(1.5,1), c(115/60,1.5), c(92/60, 1), c(33.5,35), c(17.5:17.5)) to the digested input

test_that("Testing Vanco_SS_dose_calculator function design",{
  expect_error(Vanco_SS_dose_calculator('25.0', 43.0, 1000, 8, 1, 1.5, 1, 35, 17.5), "Sorry, you need to ensure all your inputs are numeric. The currently entered inputs are character numeric numeric numeric numeric numeric numeric numeric numeric numeric") # expect_error test comparing the input error to the predicted error
})


vancomycin.calculator/R/Vancomycin_Dose_Calculator.R
#' Vancomycin Dose Calculator
#' @description The function calculates a new dosing regimen for patients receiving Vancomycin using the Sawchuk-Zaske calculation
#' strategy. This strategy determines the elimination rate constant(k), the half-life(t1/2), the volume of distribution (Vd), the
#' new dosing interval (tau), the new dosing rate recommendation (k0), the expected peak level (C 1h), and the expected trough
#' level (C tr). Any relevant equations can be found online or by viewing the equations in the function. Two or three serum
#' concentrations are needed, along with the assumption that the patient is at steady-state (receiving their 4th or 5th dose
#' depending on the half-life of the drug). Using the patient parameters, individualized dosing regimens can be created to achieve
#' ideal target levels. The goal of this function is to facilitate the dosing adjustments that are typically conducted in a
#' hospital setting, and to allow clinical pharmacists to make patient specific judgements quicker.

#' @param pre_level Patient level taken prior to providing the next dose (mg/L)
#' @param post_level Patient level taken after infusion time of dose (mg/L)
#' @param dose Amount of drug given to the patient (mg)
#' @param regimen Dosing interval currently used for the patient (Q8h, Q12h, Q18h, Q24h, Q36h, Q48h)
#' @param infusion_time Time over which the drug is infused (h)
#' @param peak_time The time when the post level is drawn (h)
#' @param trough_time The time with the pre_level is drawn (h)
#' @param peak_conc The ideal peak concentration for the patient (mg/L)
#' @param trough_conc The ideal trough concentration for the patient (mg/L)

#' @return The recommended dosing interval is returned through this function
#' @export
#'
#' @examples
#'
#' Vanco_SS_dose_calculator(11.0, 22.1, 1000, 12, 1.5, 1.33333, 55/60, 35, 17.5) # should be "This function recommends a new dose of 1500 mg given over Q 12 h. This will produce a peak of 33.15 mg/L and a trough value of 15.2692370301831 mg/L"
#' Vanco_SS_dose_calculator(25.0, 33.5, 1500, 8, 1.5, 115/60, 92/60, 33.5, 17.5) # should be "This function recommends a new dose of 1500 mg given over Q 8 h. This will produce a peak of 33.5 mg/L and a trough value of 21.5794478677179 mg/L"
#' Vanco_SS_dose_calculator(c(25,25), c(33.5, 43.0), c(1500,1000), c(8,8), c(1.5,1), c(115/60,1.5), c(92/60, 1), c(33.5,35), c(17.5:17.5)) # should be "This function recommends a new dose of 1500 750 mg given over Q 8 8 h. This will produce a peak of 33.5 32.25 mg/L and a trough value of 21.5794478677179 16.6211709853472 mg/L"
#'

Vanco_SS_dose_calculator <- function(pre_level, post_level, dose, regimen, infusion_time, peak_time, trough_time, peak_conc, trough_conc) {
  x <- c(pre_level, post_level, dose, regimen, infusion_time, peak_time, trough_time, peak_conc, trough_conc) # storing inputs
  if(!is.numeric(x)) {
    stop("Sorry, you need to ensure all your inputs are numeric. The currently entered inputs are ",
         class(pre_level), " ",
         class(post_level), " ",
         class(post_level)," ",
         class(dose)," ",
         class(regimen)," ",
         class(infusion_time)," ",
         class(peak_time)," ",
         class(trough_time)," ",
         class(peak_conc)," ",
         class(trough_conc)) # check to see if the inputs are numeric
  }

  k <- log(post_level/pre_level)*(1/(regimen-(infusion_time + peak_time +
                                                trough_time))) # calculating elimination rate constant
  t_half <- log(2)/k # calculating half life

  Vd_num <- ((dose/infusion_time)*(1-exp(-k*infusion_time))*(exp(-k*peak_time))) # numerator for Vd
  Vd_denom <- (k*post_level*(1-exp(-k*regimen))) # denominator for Vd
  Vd <- Vd_num/Vd_denom # calculating volume of distribution

  new_regimen <- log(peak_conc/trough_conc)*1/k + infusion_time + 1 # calculating new regimen
  new_regimen <- (ifelse(new_regimen<10, 8,
                         ifelse(new_regimen<15, 12,
                         ifelse(new_regimen<21, 18,
                         ifelse(new_regimen<30, 24,
                         ifelse(new_regimen<42, 36,
                         ifelse(new_regimen<48, 48 ,'Unknown'))))))) # rounding regimen to appropriate time

  rate_num <- (k*Vd*peak_conc*(1-exp(-k*new_regimen))) # dose rate numerator
  rate_denom <- ((1-exp(-k*infusion_time))*(exp(-k*peak_time))) # dose rate denominator
  rate <- rate_num/rate_denom # determining new dose rate

  new_dose <- rate*infusion_time # calculating dose from dose rate
  rounded_dose <- plyr::round_any(new_dose, 250) # rounding the dose to nearest available formulation

  new_peak <- (peak_conc*rounded_dose)/new_dose # calculating new peak level
  new_trough <- new_peak*exp(-k*(new_regimen-(infusion_time+peak_time))) # calculating new trough level
  output <-c("This function recommends a new dose of", rounded_dose, "mg", "given over Q", new_regimen, "h. This will produce a peak of",
             new_peak, "mg/L","and a trough value of", new_trough, "mg/L")
  return(paste(output,collapse=' ')) # function return
}


